components/automate-ui/src/app/pages/project/rules/project-rules.component.ts

```diff
@@ -78,48 +80,44 @@ export class ProjectRulesComponent implements OnInit, OnDestroy {
         this.store.select(updateStatus),
         this.store.select(getProjectStatus)
       ]).pipe(
-        takeUntil(this.isDestroyed),
-        map(([gStatus, uStatus, gpStatus]) => {
-          const routeId = this.route.snapshot.paramMap.get('ruleid');
-          this.isLoading =
-            routeId
-            ? (gStatus !== EntityStatus.loadingSuccess) ||
-              (uStatus === EntityStatus.loading) ||
-              (gpStatus !== EntityStatus.loadingSuccess)
-            : false;
-        })).subscribe();
+        takeUntil(this.isDestroyed)
+      ).subscribe(([gStatus, uStatus, gpStatus]) => {
+        const routeId = this.route.snapshot.paramMap.get('ruleid');
+        this.isLoading = routeId &&
+          (gStatus !== EntityStatus.loadingSuccess) ||
+          (uStatus === EntityStatus.loading) ||
+          (gpStatus !== EntityStatus.loadingSuccess);
+        });
 
       combineLatest([
         this.store.select(routeParams).pipe(pluck('id'), filter(identity)),
         this.store.select(routeParams).pipe(pluck('ruleid'), filter(identity))
       ]).pipe(
-        takeUntil(this.isDestroyed),
-        map(([project_id, rule_id]: string[]) => {
-          this.store.dispatch(new GetProject({ id: project_id }));
-          this.store.dispatch(new GetRule({
-            id: rule_id,
-            project_id: project_id
-          }));
-        })).subscribe();
+        takeUntil(this.isDestroyed)
+      ).subscribe(([project_id, rule_id]: string[]) => {
+        this.store.dispatch(new GetProject({ id: project_id }));
+        this.store.dispatch(new GetRule({
+          id: rule_id,
+          project_id
+        }));
+      });
 
       this.store.select(projectFromRoute).pipe(
         filter(identity),
-        takeUntil(this.isDestroyed),
-        map((state) => {
-          this.project = { ...state };
```

Comment:
❓ Did we **need** the copy? I didn't think so. Same with `this.rule` below -- I might be missing something, though.

Reply:
Random google result makes me think that the reducer is where you should do the copy: https://medium.com/@tomastrajan/object-assign-vs-object-spread-in-angular-ngrx-reducers-3d62ecb4a4b0

So.... let's eventually do that and leave this as-is? :wink:

Reply:
1. No worry about reducers; they were implemented from the ground up to be immutable in our code. Reducers always need to be pure functions so the ngrx store should always be immutable.
2. My concern is are we subverting that by grabbing something from the store and then manipulating it locally? Not sure, but I think that might change the ngrx store, and that would be bad. 

Even if the above is not messing with the ngrx store, though, one should always strive for immutability. Here's an interesting read on it (though not suggesting we go this full-blown route at the moment 😁 ): https://medium.com/@ezequiel/immutability-and-builders-with-typescript-b69a51c94e8c